gtk-subclass/tests/simple_application.rs

```diff
@@ -0,0 +1,227 @@
+
+            match *prop {
+                _ => unimplemented!(),
+            }
+        }
+
+        fn get_property(&self, _obj: &glib::Object, id: u32) -> Result<glib::Value, ()> {
+            let prop = &PROPERTIES[id as usize];
+
+            match *prop {
+                _ => unimplemented!(),
+            }
+        }
+    }
+    impl GApplicationImpl<Application> for SimpleApplication {
+
+        fn startup(&self, application: &Application){
+            application.parent_startup();
+
+            self.build_ui(application);
+        }
+
+
+    }
+    impl ApplicationImpl<Application> for SimpleApplication {}
+
+    pub struct SimpleApplicationStatic;
+
+    impl ImplTypeStatic<Application> for SimpleApplicationStatic {
+        fn get_name(&self) -> &str {
+            "SimpleApplication"
+        }
+
+        fn new(&self, application: &Application) -> Box<ApplicationImpl<Application>> {
+            SimpleApplication::init(application)
+        }
+
+        fn class_init(&self, klass: &mut ApplicationClass) {
+            SimpleApplication::class_init(klass);
+        }
+    }
+}
+
+
+glib_wrapper! {
+    pub struct SimpleApplication(Object<imp::SimpleApplication>):
+        [Application => InstanceStruct<Application>,
+         GApplication => InstanceStruct<GApplication>,
+         gtk::Application => gtk_ffi::GtkApplication,
+         gio::Application => gio_ffi::GApplication,
+         gio::ActionGroup => gio_ffi::GActionGroup,
+         gio::ActionMap => gio_ffi::GActionMap];
+
+    match fn {
+        get_type => || imp::SimpleApplication::get_type().to_glib(),
+     }
+ }
+
+
+impl SimpleApplication {
+    pub fn new<'a, I: Into<Option<&'a str>>>(application_id: I, flags: gio::ApplicationFlags) -> Result<SimpleApplication, glib::BoolError> {
+        use glib::object::Downcast;
+        try!(gtk::init());
+
+        unsafe {
+            match glib::Object::new(Self::static_type(), &[("application_id", &application_id.into()),
+                                                           ("flags", &flags)]){
+                Ok(obj) => Ok(obj.downcast_unchecked()),
+                Err(_) => Err(glib::BoolError("Failed to create application"))
+            }
+        }
+    }
+}
+
+// TODO: This one should probably get a macro
+impl Deref for SimpleApplication {
+    type Target = imp::SimpleApplication;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+
+            let base: Application = from_glib_borrow(self.to_glib_none().0);
+            let imp = base.get_impl();
+            let imp = imp.downcast_ref::<imp::SimpleApplication>().unwrap();
+            // Cast to a raw pointer to get us an appropriate lifetime: the compiler
+            // can't know that the lifetime of base is the same as the one of self
+            &*(imp as *const imp::SimpleApplication)
+        }
+    }
+}
+
+
+
+
+#[test]
+fn test_create() {
+    let application = SimpleApplication::new("com.github.basic",
```

Comment:
This is problematic as GTK requires a running display and requires things to be single-threaded, but Rust runs all tests on separate threads.

Reply:
So this basically needs the things mentioned for gtk-test, otherwise it won't run well